docs/guides/avxpq-multiplatform-support.md

```diff
@@ -0,0 +1,177 @@
+
+### Cross-Compilation
+
+- A technique by which you can build a binary that targets Platform A on a machine of Platform B. I.e. cross-compilation enables you to build a Windows x86_64 binary from a Linux aarch64 host.
+- Results in binaries that don't need to take the performance hit of emulation at runtime.
+
+### Host Architecture
+
+- The architecture of the CPU on which Dagger executes containers.
+
+### Target Architecture
+
+- The architecture of filesystems that Dagger expects as inputs and outputs during a given query. May or may not match the host architecture.
+
+## Usage
+
+### Setup
+
+To exec binaries on filesystems that target a different platform than that of the host, `binfmt_misc` needs to be configured on the host kernel (on the machine where Dagger will run containers).
+
+If the host is running docker, this one-liner will setup everything (only needed one time):
+
+```sh
+docker run --privileged --rm tonistiigi/binfmt --install all
+```
+
+[See the docs here for more info](https://github.com/tonistiigi/binfmt/).
+
+### Pulling Platform-specific Images
+
+Example of pulling the amd64 (a.k.a x86_64) alpine image (works even on hosts of different architectures):
+
+```graphql
+{
+  withArchitecture(architecture: "amd64") {
+    core {
+      image(ref: "alpine:latest") {
+        exec(input: { args: ["uname", "-m"] }) {
+          stdout
+        }
+      }
+    }
+  }
+}
+```
+
+### Pushing Multiplatform Images
+
+NOTE: this is a _very_ temporary solution. The better solution will involve an API type that can bundle multiple platform-specific versions of a filesystem together (see Future Changes section).
+
+```graphql
+query MultiplatformImagePush(
+  $ref: String!
+  $armFS: FSID!
+  $amdFS: FSID!
+  $s390xFS: FSID!
+) {
+  core {
+    pushMultiplatformImage(
+      ref: "localhost:5000/mymultiplatimage:latest"
+      filesystems: [$armFS, $amdFS, $s390xFS]
+    )
+  }
+}
+```
+
+### Emulated Compilation Example
+
+If `withArchitecture` is used then all core exec calls will be executed with the provided architecture, even if it doesn't match the host. E.g.
+
+```graphql
+{
+  withArchitecture(architecture: "amd64") {
+    core {
+      image(ref: "myimage:latest") {
+        exec(input: { args: ["build", "something"] }) {
+          fs {
+            id
+          }
+        }
+      }
+    }
+  }
+}
+```
+
+Will result in `build something` being called on the amd64 `myimage` even if the Dagger host is of a different architecture.
+
+[See also our test case for slightly more involved examples (the `emulateCompile` method specifically)](https://github.com/dagger/dagger/tree/cloak/core/integration/testdata/multiplatform/ext/main.go)
+
+### Cross-Compilation Example
+
+Two parts to this:
+
+1. You can obtain the current target platform via the `architecture` API
+1. You can force a query to run as the host platform by wrapping it in `withArchitecture(architecture: "host")`
```

Comment:
Does this need to be explicit? I assume running on the host platform architecture is the default anyway, can the same result not be achieved by simply omitting `withArchitecture` ?

Reply:
Yeah, the use case here is for extensions to be able to selectively use more efficient (but specialized) cross-compilation rather than slow (but universal) emulation in a way that is completely opaque+seamless to the caller.

To explain a bit more:
1. Say that I am calling an extension `builder` that builds some code and returns a filesystem with a binary. Lets say to start that `builder` just uses `exec` to compile code but has no special logic to handle cross-compilation.
2. Now I want that binary for some non-host platform, so I wrap my query in `withArchitecture(architecture: "foo")`
3. `builder` doesn't need any updates to support this thanks to emulation. All the calls `builder` makes will now be ambiently passed with the default architecture of `foo` and thus result in a filesystem of that arch.
4. *Now* let's say that `builder` is updated to support special cross-compilation. What `builder` can do is change its queries (i.e. to `exec`) to specify that the "host" architecture should always be used. It additionally can pass whatever args, env vars, etc. it needs in order to trigger cross-compilation for the target platform.
   * An example of this is in [the test case here](https://github.com/dagger/dagger/blob/540b7fe9288280afbad266bfb77450c402cd48f3/core/integration/testdata/multiplatform/ext/main.go#L68)

The end result of this is that we started with the "universal" approach of emulation, but then allowed an extension to selectively add support for faster cross-compilation (which requires specialized interaction with more "user-space" code), all without requiring the caller to change anything or be aware of any of this besides their target platform.

Allowing extensions to specify `host` as the architecture was the simplest way I found to enable this sort of use case, but obviously open to any other suggestions.

Reply:
This makes me feel like the "query wrapper" approach, although cool, is too much magic and creates a special case that all developers need to be careful to be aware of ("when making a query, all inputs are explicit, oh except ambiant architecture, that's special").

If we made `withArchitecture` a field of `Container` then the special case would disappear. It would also create a natural attachment for `push` (no need for a separate multi-platform push action if container is arch-aware).

Also I think the concept of "default architecture" is more tricky than it might appear. It's not at all obvious that one default will be correct for all use cases. I'm not sure what to do with that information, but it was very obvious in today's community call: @sipsma you mentioned the default behavior of using the buildkit worker's architecture; @aluzzardi mentioned the possibility of using the client's architecture. etc.

Reply:
> This makes me feel like the "query wrapper" approach, although cool, is too much magic and creates a special case that all developers need to be careful to be aware of ("when making a query, all inputs are explicit, oh except ambiant architecture, that's special").

That's the thing though, developers don't need to be aware of it at all. That's why the `alpine` extension worked on different platforms without me changing a single line of its code.

If extensions don't have any special logic to handle platforms, then everything will still execute and be output as the expected platform thanks to emulation. That's the universal fallback. It's only when an extension developer opts into the complication of cross-compilation that they need any awareness at all and that's 100% optional.

What you're bringing up though is extremely important: the golden rule for anything we put in ambient config in the future (i.e. cache config) is that extension developers never need to change any code to handle the config specially. They can if they want for some special purpose, but they shouldn't have to have any awareness or specialization for it.

The alternative approach of making all extensions explicitly accept platform config actually *would* force them to all be aware of it and handle it correctly, which is a lot to ask of those developers when there's an alternative option where it just all works for them automatically.

> If we made withArchitecture a field of Container then the special case would disappear. It would also create a natural attachment for push (no need for a separate multi-platform push action if container is arch-aware).

I have mixed feelings here. From my perspective, the new `Container` API is actually using the exact same pattern as we are using in the current `withArchitecture(...): Query!`. The `Container` object is a blob of state that you update and pass along through a pipeline. That's not really any different than including a blob of state as part of the `Query` object and passing that along through a pipeline; it's just one "level" higher in `Query`.

You could almost think of `Query` as a "container" of ambient config that's universally applicable to any extension? I guess? I highly suspect cache config will end up going in there too. Probably some other things too.

The tradeoff here is that if `withArchitecture` is *only* a part of `Container` then any extension that adds APIs outside of `Container` is going to have to explicitly accept optional platform config and then have special awareness of it, handle it correctly, etc. E.g. `alpine` will have to ask for a platform and then be careful pass it along to each of its subcalls to other apis. Same for every other extension.

So if we instead make `withArchitecture` a part of the global `Query` state then we can use it in `Container` (and get the nice seamless `push` like you mentioned) but also in every other extension too. 

All that being said, I do understand the appeal of being extremely explicit and limiting the magic to just the types we control (like `Container`). It's only worth having the more seamless-but-magical ambient config approach if we can truly get it fully seamless (not 95% seamelss). So far, it seems like we can achieve that with multiplatform here thanks to emulation, but it's also something we can't know for sure until users start playing with it more. If it turns out that the idea breaks down, I'm totally okay with limiting the scope of `withArchitecture` to just `Container`.

> Also I think the concept of "default architecture" is more tricky than it might appear. It's not at all obvious that one default will be correct for all use cases. 

I agree there's not an obvious default. But if we are forced to choose one, I would have to vote for the buildkit host just because it will result in faster execution (no emulation by default). But I also think @aluzzardi's suggestion that the default itself it should be configurable (possibly through a header) is good, we at minimum need to do that no matter what.